### PR TITLE
Auth providers

### PR DESCRIPTION
**Pull request by amitu**
- [x] `amitu`: implement ud function on fastn (supporting both pg and sqlite) (only to be used by processor, wasm will use ft-sdk version of ud) (not a blocker for now)
- [x] `amitu`: implement migration on fastn (supporting both pg and sqlite) #16
- [x] https://github.com/FifthTry/ft-sdk/pull/12
- [x] send email fastn 
- [x] https://github.com/FifthTry/ft-sdk/pull/15
- [x] https://github.com/FifthTry/ft-sdk/pull/19
- [ ] `ft_sdk::auth_provider::*` functions (`authenticate()`, `logout()`) (this PR)
- [ ] `ft_sdk::auth::*` functions (eg `ud()`, `is-logged-in()`) (this PR)

### CODE REVIEW

File: ft-sys/src/http.rs [hunk 1]
```diff
@@ -33,3 +33,9 @@ pub fn send_response(r: http::Response<bytes::Bytes>) {
     let (ptr, len) = ft_sys::memory::json_ptr(r);
     unsafe { http_send_response(ptr, len) }
 }
+
+
+/// doc_id: auth.fifthtry.site/create-account
```
Comment (amitu) [inline ft-sys/src/http.rs:38]:
```txt
auth.fifthtry.site/auth.wasm
id=auth.fifthtry.site/login/index.ftd
id=auth.fifthtry.site/login
id=/login.ftd
id=login

$ cat amitu.com/FASTN.ftd
-- fastn.dependency: auth-ui.fifthtry.site
provided-via: amitu.com/auth
$ cat login.ftd
-- import: auth-ui.fifthtry.site as ui
-- ui.login-page:
$
```

File: ft-sys/src/http.rs [hunk 2]
```diff
@@ -33,3 +33,9 @@ pub fn send_response(r: http::Response<bytes::Bytes>) {
...
+pub fn send_ftd(_doc_id: &str) {
```
Comment (amitu) [inline ft-sys/src/http.rs:39]:
```suggestion
pub fn send_ftd(_doc_id: &str, request_data: HashMap<String, serde_json::Value>) {
```

**Review by siddhantk232 — APPROVED** [review]

**Comment by siddhantk232** [comment]
The name `authenticate` was not making much sense to me. I've split that into `modify_current_user` (that update info of the currently logged in user), and `login`.  This splits the giant comment above `authenticate` into two pieces:

1. `login`: will do session related things. This will actually log the user in
2. `modify_current_user`: providers will use this to update the user details

both of these are WIP right now.

**Comment by siddhantk232** [comment]
I've somewhat implemented the `modify_user`, there are so many test cases for which the current code will most probably break or behave unexpectedly. This is not god, I'll start with a bunch of tests tomorrow to nail down the logic.

I've also reduced the json `data` in `fastn.user` table from:

```jsonc
{
    "<provider-name>": {
        "<provider-id>": {
            "scopes": [],  // granted scopes
            "data": {
                "UserData::VerifiedEmail": "foo@bar.com",
             }
        }
    }
}
```

to:

```jsonc
{
    "<provider-id>": {
        "scopes": [],  // granted scopes
        "data": {
            "UserData::VerifiedEmail": "foo@bar.com",
         }
    }
}
```

not sure what was the intent behind having `<provider-name>` was? Do we expect multiple entries per provider for each user, in which case?

File: ft-sdk/src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,5 @@
+/// debug print queries generated by diesel
+/// TODO: add a crate feature to conditionally enable this
```
Comment (amitu) [inline ft-sdk/src/utils.rs:2]: This should not be a doc comment, but a simple comment.

File: ft-sdk/src/auth.rs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+pub struct ProviderUserData {
+    ud: UserData,
+    user_visible_identity: String,
+    provider: String,
+}
+
+/// Get the currently logged in user's userid.
+/// If no user is logged in, an empty user will be created and its id will
+/// be returned.
```
Comment (amitu) [inline ft-sdk/src/auth.rs:37]: This feels wrong.

File: ft-sdk/src/utils.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+///     .select(user::name)
+///     .filter(user::id.eq(9))
+///     .limit(1);
+///
+/// ft_sdk::utils::dbg_query(&query);
+/// ```
+pub fn dbg_query<T: diesel::query_builder::QueryFragment<diesel::pg::Pg>>(query: &T) {
+    #[cfg(feature = "debug")]
+    ft_sdk::println!("{:?}", diesel::debug_query::<diesel::pg::Pg, _>(query));
+}
```
Comment (amitu) [inline ft-sdk/src/utils.rs:31]: We should also add a `ft_sdk::dbg!()`, which acts like acts like `dbg!()`. Actually, can we not make `println!()`, `dbg!()`, `log!()` etc use our internal helper, so we do not have to ask users to update their `println!()` calls to `ft_sdk::println!()` calls?

File: ft-sdk/src/auth_provider.rs [hunk 6]
```diff
@@ -0,0 +1,838 @@
...
+    // seen their GitHub user id. If we show that user is logged in twice via GitHub, we have to
+    // show some identity against each, and we will use this identity. Identity is mandatory. It
+    // will be stored as UserData::Identity.
+    //
+    // For the same provider_id, if identity changes, we will only keep the latest identity.
+    identity: &str,
+    data: Vec<ft_sdk::auth::UserData>,
+    user_id: Option<ft_sdk::auth::UserId>,
+    // TODO:
+    // _token: Option<serde_json::Value>,
```
Comment (amitu) [inline ft-sdk/src/auth_provider.rs:46]: Token should be stored in `ft_sdk::auth::UserData::Custom`. Problem is we do not know if there is a single token or more, like a lot of oath provider need auth token and refresh token, and both has to be stored. Custom can handle such stuff.

File: ft-sdk/src/auth_provider.rs [hunk 7]
```diff
@@ -0,0 +1,838 @@
...
+) -> Result<ft_sdk::UserId, AuthError> {
+    let user_id = if let Some(id) = user_id {
+        id
+    } else {
+        create_empty_user(conn)?
+    };
+
+    modify_user(&user_id, conn, provider_id, identity, data)?;
+
+    Ok(user_id)
```
Comment (amitu) [inline ft-sdk/src/auth_provider.rs:56]:
```suggestion
    if let Some(id) = user_id {
        modify_user(id, conn, provider_id, identity, data)
    } else {
        create_user(conn, provider_id, identity, data)?
    };
```

File: ft-sdk/src/auth_provider.rs [hunk 8]
```diff
@@ -0,0 +1,838 @@
...
+    #[error("ft_sdk::auth::UserData::Name is required")]
+    NameNotProvided,
+}
+
+/// returns `true` if there's a [UserData::VerifiedEmail] for the provided email
+///
+/// this makes a db call to check if the email is already verified.
+pub fn check_email(
+    conn: &mut ft_sdk::Connection,
+    _email: &str,
```
Comment (amitu) [inline ft-sdk/src/auth_provider.rs:72]:
```suggestion
    email: &str,
```

File: ft-sdk/src/auth_provider.rs [hunk 9]
```diff
@@ -0,0 +1,838 @@
...
+) -> Result<bool, diesel::result::Error> {
+    use db::fastn_user;
+    use diesel::dsl::{count, sql};
+    use diesel::prelude::*;
+    use diesel::sql_types::{Bool, Text};
+
+    #[cfg(not(feature = "postgres"))]
+    let filter = sql::<Bool>("data->'email'->'data'->'verified_emails' LIKE '%\"")
+        .bind::<Text, _>(_email)
+        .sql("\"%'");
```
Comment (amitu) [inline ft-sdk/src/auth_provider.rs:82]: This assumes email is only provided by an auth provider named `email`. We have to loop through all provider data and see if any one of them provided us verified email. Github, Google etc provide verified email as well for example. If you want to check for verified email provided by a given provider only we should add that providers name as an argument to this function. Maybe as `provider: Option<String>` so I can pass None, and am interested in first verified email.

File: ft-sdk/src/auth_provider.rs [hunk 10]
```diff
@@ -0,0 +1,838 @@
...
+    provider_id: &str,
+    // GitHub may use username as Identity, as user can understand their username, but have never
+    // seen their GitHub user id. If we show that user is logged in twice via GitHub, we have to
+    // show some identity against each, and we will use this identity. Identity is mandatory. It
+    // will be stored as UserData::Identity.
+    //
+    // For the same provider_id, if identity changes, we will only keep the latest identity.
+    identity: &str,
+    data: Vec<ft_sdk::auth::UserData>,
+    user_id: Option<ft_sdk::auth::UserId>,
```
Comment (amitu) [inline ft-sdk/src/auth_provider.rs:44]: Remove this?

Comment (amitu) [reply ft-sdk/src/auth_provider.rs:44]: For create account we do not need this, but for login we do.

**Pull request merged**